### ISSUE TITLE
`use_self` lint triggers on #[derive(Deserialize)]

### ISSUE THREAD

**Issue by rossmacarthur** [rust-lang/rust-clippy#6894]
Lint name: [`use_self`](https://rust-lang.github.io/rust-clippy/master/index.html#use_self)

I tried this code:

```rust
// lib.rs
use serde::Deserialize;

#[derive(Deserialize)]
struct Settings {}
```

And ran

```
cargo +nightly clippy -- -D clippy::use-self
```

I expected to see this happen: *no error*

Instead, this happened: *`use_self` lint triggers*

```
error: unnecessary structure name repetition
 --> src/lib.rs:4:8
  |
4 | struct Settings {}
  |        ^^^^^^^^ help: use the applicable keyword: `Self`
  |
  = note: requested on the command line with `-D clippy::use-self`
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#use_self

error: aborting due to previous error

error: could not compile `clippy-test`

To learn more, run the command again with --verbose.
```

### Meta

- `cargo +nightly clippy -V`: 
   ```
   clippy 0.1.52 (b3e19a22 2021-03-12)
   ```
- `rustc +nightly -Vv`:
   ```
   rustc 1.52.0-nightly (b3e19a221 2021-03-12)
   binary: rustc
   commit-hash: b3e19a221e63dcffdef87e12eadf1f36a8b90295
   commit-date: 2021-03-12
   host: x86_64-apple-darwin
   release: 1.52.0-nightly
   LLVM version: 12.0.0
   ```

**Comment by Y-Nak** [comment]
This seems a duplicate of #6818.

**Comment by rossmacarthur** [comment]
You're right
